feat: add system.unicode

wxybear opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes into CHANGELOG.md):
Add `system.unicode` table, containing a list of Unicode characters and their properties. Closes #80055.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

wxybear committed:
feat: init commit, add system.unicode

wxybear committed:
Merge remote-tracking branch 'origin/master'

wxybear committed:
feat: add tests for system.unicode and some refine

wxybear committed:
feat: add system.unicode docs and refine

wxybear committed:
comments: refine

wxybear committed:
style: refine

UnamedRus commented on docs/en/operations/system-tables/unicode.md:
```diff
@@ -0,0 +1,164 @@
+──────
+code_point:                      a
+code_point_value:                97
+Alphabetic:                      1
```
ClickHouse is using snake case for system table columns.

wxybear replied:
Yes! I didn’t use snake case in some columns because, in addition to code_point, code_point_value and I want to add a notation column, the other columns are property columns of unicode code points, which are defined in ICU lib and are also the information returned by the API. Of course, I can also do a tolower. If the result returned by the API is kept, one advantage is that users can query the corresponding ICU documents. What do you think?

wxybear committed:
feat: add notation column and some refine

wxybear committed:
style: remove whitespaces

wxybear committed:
style: trim whitespace

wxybear committed:
Merge branch 'ClickHouse:master' into feat_system_unicode

wxybear committed:
feat: converted property names to snake case

wxybear committed:
Merge remote-tracking branch 'origin/master' into feat_system_unicode

wxybear committed:
fix: bug after merge

alexey-milovidov commented:
Fast test has most of the libraries disabled, including ICU. You can disable the test under Fast test.

wxybear committed:
feat: disable system unicode test when running fast test

wxybear committed:
fix: wrong char_name_buffer size

wxybear committed:
style: trim whitespace

wxybear commented:
It looks like there are two small issues. I will check if one of them is an issue with icu lib. I will complete it within a week.

wxybear committed:
feat: add some missing properties

wxybear committed:
style: remove "auto _"

wxybear committed:
fix: msan and md

wxybear committed:
fix: system.unicode have no comments

wxybear committed:
Merge remote-tracking branch 'origin/master' into feat_system_unicode

wxybear commented:
@alexey-milovidov hi, do you know why does the new tests take so long to run? Even the debug binary does not take so long in my local machine, Thanks!
https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80857&sha=cdb6f846c272ff26f49a1347593d7e79861cf7da&name_0=PR&name_1=Stateless+tests+%28amd_asan%2C+flaky+check%29
> 2025-06-14 21:43:44 Reason: Test runs too long (> 180s). Make it faster.

alexey-milovidov commented:
We do artificial slow-downs by inserting random sleeps at synchronization points - it is named Thread Fuzzer.

maxknv committed:
ping ci

wxybear commented:
@alexey-milovidov In icu4c, some properties are indeed obtained using data structures protected by global mutex, such as gCommonICUDataArray, gCommonCleanupFunctions, etc. Can I also disable tsan?

wxybear committed:
feat: disable 03521_system_unicode.sql flaky test

wxybear committed:
Merge remote-tracking branch 'origin1/feat_system_unicode' into feat_system_unicode

wxybear committed:
feat: system_unicode use predicate and restore flaky test

wxybear committed:
style: trim whitespaces

Algunenano committed:
Merge branch 'master' into feat_system_unicode

wxybear committed:
Merge remote-tracking branch 'origin/master' into feat_system_unicode

wxybear committed:
Merge remote-tracking branch 'origin1/feat_system_unicode' into feat_system_unicode

wxybear committed:
style: standardize include format

alexey-milovidov committed:
Merge branch 'master' into feat_system_unicode

alexey-milovidov commented:
It is quite slow, runs in 4 seconds on my machine.

alexey-milovidov committed:
Speed-up 5 times

alexey-milovidov committed:
Fix bad code

alexey-milovidov commented:
It has a bug:

```
code_point:                      🍆
code_point_value:                127814
notation:                        U+0005
```

Moreover, I can't find the description "eggplant" in the output.

alexey-milovidov commented:
Found, it is AUBERGINE.

alexey-milovidov committed:
Fix error

alexey-milovidov commented:
The code does not check errors.

alexey-milovidov committed:
Faster

wxybear commented:
@alexey-milovidov Thank you very much for your commits!! Originally, I assumed the unhandled errors stemmed from our server-side generated dataset, which should be intrinsically valid—so some error would imply the data itself was out of place. However, I’ll add comprehensive error checks everywhere, which is safer.

wxybear committed:
feat: refine system unicode
1. use array to list props, not use prop limit
2. define U_HIDE_DEPRECATED_API to shadow depreciated **_limit enum and
   some other enum

wxybear committed:
docs: add identifier_type and remove some depreciated columns

wxybear committed:
style: format and adjust comments

wxybear committed:
Merge remote-tracking branch 'origin/master' into feat_system_unicode

alexey-milovidov commented:
It's ok now, but we need to find a way to improve the performance about 50 times.

wxybear commented:
@alexey-milovidov Sure, let me do some profiling and optimization first

wxybear committed:
feat: other properties use array(lowcardinality(string))

alexey-milovidov commented:
I think the reason for the low performance is that we use separate methods to access every single property, and these methods lock a mutex inside for each invocation.

I'm comparing with this tool https://github.com/arp242/uni, and it has a much better performance.

alexey-milovidov committed:
Merge branch 'master' into feat_system_unicode

wxybear committed:
Merge remote-tracking branch 'origin/master' into feat_system_unicode

wxybear commented:
In my opinion, the time we are currently spending is not due to locking a mutex, but mainly due to the need to initialize the data?

I looked at the uni implementation, which generates a go file and maps props directly to the data in the map in the .go code. Can we also consider a similar way to generate cpp files through shell scripts (such as gen.sh)

wxybear committed:
feat: iterate by props (speed up 1.5x)

wxybear committed:
Merge remote-tracking branch 'origin1/feat_system_unicode' into feat_system_unicode

wxybear committed:
docs: refine unicode.md

wxybear committed:
feat: only read columns we need when read system.unicode

wxybear committed:
Merge remote-tracking branch 'origin/master' into feat_system_unicode

wxybear committed:
Merge remote-tracking branch 'origin/master' into feat_system_unicode

alexey-milovidov committed:
Merge branch 'master' into feat_system_unicode

alexey-milovidov committed:
Merge branch 'master' into feat_system_unicode-2

alexey-milovidov committed:
Merge branch 'master' into feat_system_unicode

alexey-milovidov committed:
Merge branch 'master' into feat_system_unicode

alexey-milovidov committed:
Add a test

alexey-milovidov committed:
Let's look at the currently developed feature, which is the new table, `system.unicode`. I found it inconvenient that the values of numeric properties are represented by numbers:
  ```
  bidi_class:                      10
  block:                           205
  canonical_combining_class:       0
  decomposition_type:              0
  east_asian_width:                5
  general_category:                27
  joining_group:                   0
  joining_type:                    0
  line_break:                      14
  ```

  Let's replace numeric values with the corresponding enums.
  Also, let's use Bool type whenever appropriate.

  We can do it this way: for each of `int_properties`, obtain `u_getIntPropertyMinValue` and `u_getIntPropertyMaxValue`. In the constructor of the storage, obtain `u_getPropertyValueName`
  with `U_LONG_PROPERTY_NAME` and construct the Enum data type from them.

alexey-milovidov committed:
Fix docs

alexey-milovidov committed:
Also, update the docs after the recent changes to Enum and Bool

alexey-milovidov commented:
@wxybear, thanks for the amazing feature!
I presented it as the first feature of 25.11: https://presentations.clickhouse.com/2025-release-25.11/

clickgapai commented:
Hi @wxybear @alexey-milovidov — while reviewing this PR I found the following:

- **Bug (P2):** `titlecase_mapping` uses simple `u_totitle` instead of full titlecase mapping, making it identical to `simple_titlecase_mapping` — https://github.com/ClickHouse/ClickHouse/issues/103051

Happy to discuss — close anything that's wrong or already addressed.
